pkg/interop/client.go

```diff
@@ -0,0 +1,486 @@
+			return nil, err
+		}
+		keyPEM, err := ioutil.ReadAll(keyFile)
+		if err != nil {
+			return nil, err
+		}
+		cert, err := tls.X509KeyPair(certPEM, keyPEM)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+	return &tls.Config{
+		RootCAs:      rootCAs,
+		Certificates: certs,
+	}, nil
+}
+
+func decodeYAML(r io.Reader, v interface{}) error {
+	dec := yaml.NewDecoder(r)
+	dec.SetStrict(true)
+	return dec.Decode(v)
+}
+
+func NewClient(ctx context.Context, conf ClientConfig, fallbackTLS *tls.Config) (*Client, error) {
+	u, err := url.Parse(conf.ConfigURI)
+	if err != nil {
+		return nil, err
+	}
+
+	var fs afero.Fs
+	switch s := u.Scheme; s {
+	case "file":
+		fs = afero.NewOsFs()
+	default:
+		return nil, errUnsupportedURIScheme.WithAttributes("scheme", s)
+	}
+	confFile, err := fs.Open(u.Path)
+	if err != nil {
+		return nil, err
+	}
+	fs = afero.NewCacheOnReadFs(fs, afero.NewMemMapFs(), conf.CacheTime)
+
+	var yamlConf struct {
+		JoinServers []struct {
+			File    string              `yaml:"file"`
+			JoinEUI []types.EUI64Prefix `yaml:"join-eui"`
+		} `yaml:"join-servers"`
+	}
+	if err := decodeYAML(confFile, &yamlConf); err != nil {
+		return nil, err
+	}
+
+	jss := make([]prefixJoinServerClient, 0, len(yamlConf.JoinServers))
+	for _, jsConf := range yamlConf.JoinServers {
+		jsFile, err := fs.Open(jsConf.File)
+		if err != nil {
+			return nil, err
+		}
+
+		var yamlJSConf struct {
+			DNS      string             `yaml:"dns"`
+			FQDN     string             `yaml:"fqdn"`
+			Path     string             `yaml:"path"`
+			Port     uint32             `yaml:"port"`
+			Protocol JoinServerProtocol `yaml:"protocol"`
+			Headers  map[string]string  `yaml:"headers"`
+			TLS      tlsConfig          `yaml:"tls"`
+		}
+		if err := decodeYAML(jsFile, &yamlJSConf); err != nil {
+			return nil, err
+		}
+
+		var js joinServerClient
+		switch yamlJSConf.Protocol {
+		case LoRaWANJoinServerProtocol1_0, LoRaWANJoinServerProtocol1_1:
+			tlsConf := fallbackTLS
+			if yamlJSConf.TLS.IsZero() {
+				tlsConf, err = yamlJSConf.TLS.TLSConfig(fs)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			var tr *http.Transport
+			if tlsConf != nil {
+				tr = &http.Transport{
+					TLSClientConfig: tlsConf,
+				}
+			}
+			js = &joinServerHTTPClient{
+				Client: http.Client{
+					Transport: tr,
+				},
+				NewRequestFunc: makeJoinServerHTTPRequestFunc("https", yamlJSConf.DNS, yamlJSConf.FQDN, yamlJSConf.Path, yamlJSConf.Port, yamlJSConf.Headers),
+				Protocol:       yamlJSConf.Protocol,
+			}
+		default:
+			return nil, errUnknownProtocol
+		}
+		for _, pre := range jsConf.JoinEUI {
+			jss = append(jss, prefixJoinServerClient{
+				joinServerClient: js,
+				prefix:           pre,
+			})
+		}
+	}
+	sort.Slice(jss, func(i, j int) bool {
+		pi, pj := jss[i].prefix, jss[j].prefix
+		if pi.Length != pj.Length {
+			return pi.Length > pj.Length
+		}
+		return pi.EUI64.MarshalNumber() > pj.EUI64.MarshalNumber()
```

Comment:
I would do this one ascending and the bit prefix descending

Reply:
Why?